lib/Parse/ParseType.cpp

```diff
@@ -968,152 +1025,192 @@ ParserResult<TupleTypeRepr> Parser::parseTypeTupleBody() {
       // If there were two names, complain.
-      if (element.NameLoc.isValid() && element.SecondNameLoc.isValid()) {
-        auto diag = diagnose(element.NameLoc, diag::tuple_type_multiple_labels);
-        if (element.Name.empty()) {
-          diag.fixItRemoveChars(element.NameLoc,
-                                element.Type->getStartLoc());
+      if (NameLoc.isValid() && SecondNameLoc.isValid()) {
+        auto Diag = diagnose(NameLoc, diag::tuple_type_multiple_labels);
+        auto Name = Element.getDeferredName();
+        auto NameText = SourceMgr.extractText(
+            Name->getRaw().getDeferredTokenRangeWithoutBackticks(),
+            L->getBufferID());
+        if (NameText == "_") {
+          Diag.fixItRemoveChars(NameLoc, TypeLoc);
         } else {
-          diag.fixItRemove(
-            SourceRange(Lexer::getLocForEndOfToken(SourceMgr, element.NameLoc),
-                        element.SecondNameLoc));
+          Diag.fixItRemove(SourceRange(
+              Lexer::getLocForEndOfToken(SourceMgr, NameLoc), SecondNameLoc));
         }
       }
-      continue;
-    }
-
-    // If there was a first name, complain; arguments in function types are
-    // always unlabeled.
-    if (element.NameLoc.isValid() && !element.Name.empty()) {
-      auto diag = diagnose(element.NameLoc, diag::function_type_argument_label,
-                           element.Name);
-      if (element.SecondNameLoc.isInvalid())
-        diag.fixItInsert(element.NameLoc, "_ ");
-      else if (element.SecondName.empty())
-        diag.fixItRemoveChars(element.NameLoc,
-                              element.Type->getStartLoc());
-      else
-        diag.fixItReplace(SourceRange(element.NameLoc), "_");
     }
-
-    if (element.SecondNameLoc.isValid()) {
-      // Form the named parameter type representation.
-      element.UnderscoreLoc = element.NameLoc;
-      element.Name = element.SecondName;
-      element.NameLoc = element.SecondNameLoc;
+  } else {
+    for (unsigned i = 0; i < Elements.size(); i++) {
+      // If there was a first name, complain; arguments in function types are
+      // always unlabeled.
+      auto Element = Elements[i];
+      SourceLoc NameLoc, SecondNameLoc, TypeLoc;
+      std::tie(NameLoc, SecondNameLoc, TypeLoc) = ElementsLoc[i];
+      if (NameLoc.isValid()) {
+        auto NameText = GetNameText(Element.getDeferredName());
+        if (NameText != "_") {
+          auto NameIdentifier = Context.getIdentifier(NameText);
+          auto Diag = diagnose(NameLoc, diag::function_type_argument_label,
+                               NameIdentifier);
+          auto SecondNameText = GetNameText(Element.getDeferredSecondName());
+          if (SecondNameLoc.isInvalid())
+            Diag.fixItInsert(NameLoc, "_ ");
+          else if (SecondNameText == "_")
+            Diag.fixItRemoveChars(NameLoc, TypeLoc);
+          else
+            Diag.fixItReplace(SourceRange(NameLoc), "_");
+        }
+      }
```

Comment:
It it possible to move these diagnostics to `ASTGen`? As far as I understand, `DeferringContextRAII` is used only for this. If we can move this diagnostics to `ASTGen`, we can separate this diagnostics to `generate(TupleTypeSyntax)` and `generate(FunctionTypeSyntax)`.

Reply:
Yes, your understanding is correct and this could indeed be moved into `ASTGen`.
However, I feel like that would contradict our goal of keeping syntax related operations in the `Parser`.
I think `ASTGen` should not be responsible for syntax diagnostics. It should do one operation only: transform `libSyntax` to AST.